dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -305,21 +306,55 @@ private RetryingCallback(
             return Futures.immediateFuture(response);
         }
 
-        @SuppressWarnings("FutureReturnValueIgnored") // error-prone bug
+        @SuppressWarnings({"FutureReturnValueIgnored", "CheckReturnValue"})
         private ListenableFuture<Response> scheduleRetry(Meter meter, long backoffNanoseconds) {
             meter.mark();
             if (backoffNanoseconds <= 0) {
                 return wrap(delegate.execute(request));
             }
             DetachedSpan backoffSpan = span.childDetachedSpan("retry-backoff");
-            ListenableScheduledFuture<ListenableFuture<Response>> scheduled = scheduler.schedule(
+            // Code beyond this line implements the following without risking leaking responses:
+            // ListenableScheduledFuture<ListenableFuture<Response>> scheduled = scheduler.schedule(
+            //     () -> {
+            //         backoffSpan.complete(RetryingCallbackTranslator.INSTANCE, this);
+            //         return delegate.execute(request);
+            //     }, backoffNanoseconds, TimeUnit.NANOSECONDS);
+            // return wrap(DialogueFutures.transformAsync(scheduled, input -> input));
+            SettableFuture<Response> responseFuture = SettableFuture.create();
+            // Scheduler future must not be cancelled, otherwise responses will leak.
+            scheduler.schedule(
                     () -> {
                         backoffSpan.complete(RetryingCallbackTranslator.INSTANCE, this);
-                        return delegate.execute(request);
+                        if (responseFuture.isDone()) {
```

Comment:
Unfortunately not that I'm aware of because the failure was the result of a race. I was able to put together an external reproducer which caused the leak every 30,000 requests or so, but not the sort of thing that would make a reasonable unit test.